Remove single install feature

gabyx opened the pull request:
This removes the single repo install stuff.
- `install.sh --single` is now a install into one repository (no more the notation of single install as before)
- `githooks.sh` instead of base_tempalte.sh
- `githooks-symlink.sh` for deploying
- `update_release_repo`: fetching and merging for updating, fetching only touches the remote branch, fast-forward merging does the update.
- registering mechanism is kept but now at `$INSTALL_DIR/registered`. We need this more then ever, for proper uninstall.

**Legacy Fixing**
- There is a `install.sh::apply_legacy_fixes` function which can be used to implement (without any dependencies) legacy fixes: There, for `LEGACY_FIXTURE_SINGLE_INSTALL`, we need to enter the version number on the latest master **before** we merge this PR. This function does some cleanup. I still have to test this path, it seems to work however. Prints also some **MIGRATION!!** warning to the user which he **has to confirm**.
  - Running through the `registered` (so far only non-single installs) list and updates the hooks right away... If there is no registered file, the best we can do is print a migration warning...
There is also `install.sh::apply_legacy_cleanup` which we can use to clean up stuff in the future. of course its not nice, but thats how it is with legacy versions...

It seems to work fine.
Tests should work. Still need to check...

gabyx commented:
I dont understand why test-arch.sh is failing on travis. Locally it works. Some permissions strangeness?? @rycus86 Do you have a clue??

rycus86 commented:
Hm, weird. Do you run the whole Arch test locally? Seems to be these two are failing:
```
Failed tests: 
- /var/lib/tests/step-103.sh (1 -- ! Expected shared hooks to be installed. 'ls: cannot access '/root/.githooks/shared': Operation not permitted')
- /var/lib/tests/step-109.sh (1 -- ! Expected only server hooks to be installed (0))
```

What are those two tests doing? Anything in there you could think of going wrong?
Btw, coverage also has two failing tests but different ones. :(
```
step-042.sh (1 -- ! Could not reset origin/master to trigger update.)
```

Not sure which one is the other failure unfortunately, kcov eats some of the output I suspect.

gabyx commented:
Coverage is a pain in the ass: kcov has some issues (using ubuntu 20.04 with docker) since in later versions >v34. coverage output is always 0, I tried to fix the coverage docker image. so for without any luck. Not even locally the test run with >v33. v33 has some old git I usuppose. But `git reset --hard HEAD^` should work (that the test which is failing in coverage). Strange. I try to address arch tests first. I have no idea... :-) Lets see..

rycus86 commented:
I tried to update kcov once but gave up, too much has changed between 33 and 34. :(

matthijskooijman commented on install.sh:
```diff
@@ -88,6 +85,96 @@ execute_installation() {
+    echo "!! Because we cannot locate these installs, "
+    echo "!! you should use the \"install into existing repositories\""
+    echo "!! later on during this install procedure to reinstall into found"
+    echo "!! single install repositories for a clean up."
```
This might be a bit confusing wording. If we cannot locate them, then how can we later fix them anyway? I think we *can* locatie them (by searching), but have no predefined list. So maybe "Because these installs were not kept in a list," and maybe "to search the filesytem for these repositories for a clean up"?

Also, I'm a bit confused about the meaning of `$WHAT` here. If there is no `autoupdate/registered` list, what does that mean? Isn't it just that non of the installed repos were marked for autoupdate? So why should this message change?

Finally: There *might* not be any leftover repositories, right? If all repos had autoupdate enabled, they would be fixed now. So maybe the message should clarify that?

matthijskooijman replied:
Also: the install into existing repositories might not actually run later (e.g. when run with `--single`, or `DO_UPDATE_ONLY`)?

gabyx replied:
`autoupdate/registered` list unfortunately only contained **non-single** installs -> thats why we can do the most appropriate fix.
Good point: Maybe I should implement a search similar to the  `install into existing repos` right here
In both cases the user needs a search path to enter to search all repos to make **no leftover**

matthijskooijman replied:
Might be better to just tell the user a command they can run to search later?

gabyx replied:
true

matthijskooijman commented on install.sh:
```diff
@@ -88,6 +85,96 @@ execute_installation() {
+    echo "!!"
+    echo "!! All still existing $WHAT install repositories:"
+    echo "!!   - \`install.sh --single\` or with "
+    echo "!!   - \`git hooks install\`"
```
Should this say "...repositories that were installed using:"? It's a bit unclear what thist list means now.

gabyx replied:
thx, should be 
```
echo "!! All still existing $WHAT install repositories created by either:"
```

matthijskooijman commented on install.sh:
```diff
@@ -88,6 +85,96 @@ execute_installation() {
+#
+# Returns:  0
+############################################################
+apply_legacy_cleanup() {
```
Why is this a separate function? Why not just call this `git config` line directly from `apply_legacy_fix_single_install` for each repo?

gabyx replied:
Good point. Change that. LegacyFixes should be self contained as much as possible...

matthijskooijman replied:
Exactly.

matthijskooijman commented on install.sh:
```diff
@@ -261,7 +348,22 @@ should_skip_install_into_existing_repositories() {
+
+############################################################
+# Check if the install script is
+#   running in for a single repository without templates.
```
This documentation (for `is_reset_all_standalone`) seems a copy-paste from `is_single_install`.

gabyx replied:
Jup, leftover, remove it, thx!

gabyx replied:
Not used!

matthijskooijman commented on install.sh:
```diff
@@ -88,6 +85,96 @@ execute_installation() {
+        WHAT="single & non-single"
+    fi
+
+    echo "!! MIGRATION INFORMATION: From now on single installs are deprecated: "
```
This says single installs are deprecated, but `--single` is still a supported option. Isn't the thing more that `--single` now no longer implies no autoupdate, but still installs just into a single repo (as opossed to searching the filesystem for all repos)?

gabyx replied:
Exaclty --single is now meaning "one install into one repo" nothing more. Which is supported by the cli.sh.

The meaning was twofold before: --single - install into 1 repo, and no updates!

matthijskooijman replied:
So then the message should not say "single installs are deprecated", I think?

rycus86 commented:
Thanks for the PR @gabyx and for the review @matthijskooijman !

This is pretty massive, so I'll probably only be able to look at it properly over the weekend.
I wanted to ask though, can we not break this up into multiple changes so there isn't so much changing at each step, and we could actually properly reason about and review each of those changes? Something like:

1. Show deprecation warning when running hooks in existing single install repos, also make new single installs fail with a message that it's unsupported
2. Get the cli/aliased script run from the release checkout
3. Generate a new template folder to use for init.templateDir and core.hooksPath target somewhere around (maybe inside) the release checkout and rip out the logic to try and find this place during install -- maybe at this point we actually have to have the generated wrappers already so we don't get a dirty checkout, I'm not sure
4. Use wrappers instead of copies of the base template
5. All the other things I forgot...

rycus86 commented:
Forgot to add: with those steps being one PR each.

gabyx commented:
> I tried to update kcov once but gave up, too much has changed between 33 and 34. :(

The update is not the problem, the bash execution is :-)
I can send the converage.sh docker changes in a gist for you to test...

gabyx commented:
Steps 1 is good! Need to do that still.
Steps 2-4 are I think already done. I dont know exactly what you mwan with 3.
@rycus86 : Be awarem the change look big, but are not. 
There is not much logic change at least for the templateFolder and such.
we still have 
```
   .githooks/
      - release/ --> same
      - bin/  --> should be deleted, not needed anymore (need to check that)
      - shared/ --> same
     -  templates/ --> all files are the githooks-symlink.sh wrapper
      - registerd -> registration mechanism (all repos using githooks and not using core.hooksPath are listed here, might be beneficial later on!, its not used anymore yet...)

Ist mostly renaming. (I could have done this maybe, in a seperate PR, but ok here it is)

gabyx commented on githooks-symlink.sh:
```diff
@@ -0,0 +1,24 @@
+# Version: 2005.132251-e7d7bd
+
+INSTALL_DIR=$(git config --global githooks.installDir)
+GITHOOKS_SCRIPT="$INSTALL_DIR/release/githooks.sh"
```
`release` is the git checkout. We already have this change in `master`FYI :-)

matthijskooijman replied:
Yeah, I would have expected `$INSTALL_DIR` to *be* the checkout, but that is probably tricky with migrations. After reading the reset of the diff, I think I understand better, so I guess this is ok now (but I don't see a resolve conversion button here, hm.)

gabyx commented on install.sh:
```diff
@@ -46,19 +46,17 @@ execute_installation() {
-        prepare_target_template_directory || return 1
+    # Apply legacy fixes
+    if ! is_update_only; then
+        apply_legacy_fixes || return 1
```
This is an error! Thx. Should be done always. No excuse :-)

matthijskooijman replied:
Yeah, agreed.

gabyx commented on install.sh:
```diff
@@ -88,6 +85,96 @@ execute_installation() {
+    LEGACY_FIXTURE_SINGLE_INSTALL=$(echo "$INSTALLED_VERSION 2004.272234-8af36c" | awk "{ print (\$1 <= \$2) }")
+
+    if [ "$LEGACY_FIXTURE_SINGLE_INSTALL" = "1" ] &&
+        LEGACY_FIXTURE_SINGLE_INSTALL="true"; then
```
I dont want bool variables with "0" or "1", I set it to "true" and always test on its "true" value later...

matthijskooijman replied:
Then why not put the `="true"` after the `then` instead of in the `if`? That would make it a lot more readable. Also, maybe just use two different variable names?

gabyx replied:
Rewrite it, looks ugly.

gabyx commented on githooks.sh:
```diff
@@ -891,26 +918,26 @@ update_release_clone() {
             echo "  See \`git hooks config [set|print] update-clone-url\` and" >&2
             echo "      \`git hooks config [set|print] update-clone-branch\`" >&2
-            echo "  Either fix this or delete the clone \`$CLONE_DIR\` to trigger" >&2
+            echo "  Either fix this or delete the clone \`$GITHOOKS_CLONE_DIR\` to trigger" >&2
```
No scripts outside the clone dir. This githooks.sh is running and will get removed, which is possible because the shell interpreted has already sourced this whole script under linux this does not pose a problem so far, we rely on the behavior that we can rm running scripts (headache)... on windows I hope it does not break. it should not with git-bash...

matthijskooijman replied:
Huh? But this echo instructs the user to remove the directory, right? Suggesting that that should subsequently trigger a new clone? But when the user deletes that directory, no scripts are running, and on the next git push or whatever, the hooks will just fail because  `githooks.sh` was deleted?

gabyx commented on githooks.sh:
```diff
@@ -871,15 +867,46 @@ is_git_repo() {
+
+    GITHOOKS_CLONE_DIR="$INSTALL_DIR/release"
+    GITHOOKS_CLONE_DIR_IS_UPDATED="false"
     PULL_ONLY="true"
```
I ve refactored the whole update stuff into the install.sh (see my update commits)
Its bad to merge the stuff inside githooks.sh. githooks.sh is only allowed to fetch changes see if updates happened and then dispatch to the install.sh which does the update of the clone, then restarts itself

matthijskooijman replied:
That sounds better, yeah.

gabyx commented on githooks.sh:
```diff
@@ -919,76 +946,76 @@ update_release_clone() {
             return 1
         fi
+
+        GITHOOKS_RELEASE_COMMIT=$(execute_git "$GITHOOKS_CLONE_DIR" rev-parse "$GITHOOKS_CLONE_BRANCH")
```
no explizit is better here. Refactored ths update shizzle... in my next commits

matthijskooijman replied:
But is using the branch name more explicit? It seems like a roundabout way to express what you mean (which is, I think, is the *currently checked out version*, which is HEAD, different from the fetched version). But if you refactored this, this might no longer be relevant :-)

gabyx commented on githooks.sh:
```diff
@@ -1016,16 +1055,25 @@ should_run_update() {
+        return 1
     fi
 
+    LATEST_VERSION=$(grep -E "^# Version: .*" <"$INSTALL_SCRIPT" | head -1 | sed -E "s/^# Version: //")
```
Thats my problem. After having only a clone repo inside the install folder. How do we know the version, we know the version after fetch of the latest commit and the one we would update. Its quite tricky

**Here** its hopefully only user output, nothing more, I try to only use git SHA....

matthijskooijman replied:
I'm not sure I understand what you're saying here. This particular line prints the *current* version, right? So you could ask git to describe the current version (in the same way to post-commit hook generates the version number)? Or we could let the post-commit hook also generate a `VERSION` file, that is easier to read than grepping through the install script.

gabyx commented on githooks.sh:
```diff
@@ -1016,16 +1055,25 @@ should_run_update() {
+    LATEST_VERSION=$(grep -E "^# Version: .*" <"$INSTALL_SCRIPT" | head -1 | sed -E "s/^# Version: //")
+    echo "Githooks install script now at version: $LATEST_VERSION"
+
+    if DO_UPDATE_ONLY="true" sh <"$INSTALL_SCRIPT"; then
```
I changed now everything to arguments, its ugly to use ENVs... The other thing about directly running, I need to postpone till the end...changing that might be fragile as well :-) dont know... oyur probably right

matthijskooijman replied:
I later had another thought: Maybe this should also pass the *current* version as a commandline argument to (something like `"$INSTALL_SCRIPT" --after-auto-update "$VERSION"`), which again removes the need for the `install.sh` to figure out the old version (which I think it now does on the copied cli.sh file?) for deciding whether to run legacy stuff.

matthijskooijman commented:
I think that putting the `s/yes/true/` and `s/base-template.sh/githooks.sh/` into their own commits would already significantly clean up things. If you did those with a big search/replace, those should be easy: Just go back to master, do each of the search/replaces and commit them separately, then do `git reset original_sha .` to reset the index to the state the PR is in now, then commit that as a third commit (and run `git checkout .` to match the working copy with the index again). Then you'll have two commits with the two replaces, leaving one commit with the actual changes.

It might be good to pick out more changes from that last commit, but that is then already easier.

gabyx commented:
This is true. I made some greate cleanup right now. I probably need to make stage all regex replaces to a seperate commit, but dont have time right now. Maybe you can go look over the latest changes.
It condensed again. Tests alpine, alpine-user work so far.

gabyx commented:
Implemented some git logic to detect if we need to apple a legacy fix. This is done rigth after we updated the release clone.

**Note:** ONLY `install.sh` applies updates (merging of the fetched branch) of the release clone anymore. `cli.sh` and `githooks.sh` are only allowed to just fetch changes or newly-clone (in case a release clone is missing -> mostly for tests...)

rycus86 commented:
> Implemented some git logic to detect if we need to apply a legacy fix. This is done rigth after we updated the release clone. This needs a proper review, and we need to test this legacy patch stuff.

Not really sure how this relates to deprecating/removing single installs. Again, I'd recommend we don't sink any more time into this big PR that combines a few changes together, and instead break it up into smaller pieces so it's easier to reason about code and test changes. Thanks!

gabyx commented:
I removed the renaming from the PR. This can be done at last.

gabyx commented:
@rycus86 : I would love to have simple steps here. But I could not come up with different PR because of the following:

>Show deprecation warning when running hooks in existing single install repos, also make new single installs fail with a message that it's unsupported

First: As soon as such a PRs Install script gets fetched (global install), we should make sure that there are no single install repos anymore. We have no way of detecting single installs, they have their own script and will continue to run (with our now updated install dir, which is kind of ugly, so they continue to run based on a newer version of the install folder, of course this will work as we did not change lots of stuff inside, the structure is the same). Since we cannot update these hooks, we can also not show a deprecation warning when running such a hook.
Second: We can however show a deprecation warning when ever this PRs install script is fetched an we are inside an old single install repo and do a single install `--single`. **and** we would use any command from the `cli.h`. I try to make such a PR, from this one only first if its feasible and then add the stuff of the other points below:

> Get the cli/aliased script run from the release checkout

That we could move out in a seperate PR I think, but its a small change and already included in this PR...

> Generate a new template folder to use for init.templateDir and core.hooksPath target somewhere around (maybe inside) the release checkout and rip out the logic to try and find this place during install -- maybe at this point we actually have to have the generated wrappers already so we don't get a dirty checkout, I'm not sure
Use wrappers instead of copies of the base template

This is already done, in this PR we use the same `.githooks/templates` folder and just place checked in symlink script `githooks-symlink.sh` for each hook.

rycus86 commented:
Ah yes, good point on the already existing single install repos, we can't change the hooks that are already in there easily. We should fail new installs though with a clear deprecation message, and updates in a single install repo should tell users how to fix it, for example use `git hooks config something` to mark it as non-single repo (I think we have this already), then retry the update.

If the other changes are small anyway, that would be all the more reason to have them in separate PRs, please. I'm particularly interested in the change to symlinks as a separate PR, and thought a smaller PR for the cli change could also be beneficial to keep it focused and easy to review.
